### PR TITLE
Basic UI with long polling update: management of dynamic widget visibility

### LINKED ISSUE #4558: Basic UI with Edge: change of visibility not dynamically taken into account
Using Microsoft Edge, the change of visibility of a sitemap element is not dynamically taken into account. You have to reopen the current page.
Of course, it is working in Firefox or Chrome.

### PR DESCRIPTION
**Pull request by lolodomo**
Fixes #4558

Cleanup code for the long polling update method with dataModelLegacy suppressed
and common code created in AbstractChangeListener

Signed-off-(by: Laurent Garnier <lg.hc@free.fr>

### CODE REVIEW

**Comment by lolodomo** [comment]
It is only in WIP step only because I don't want an immediate merge and introduce a conflict with antoher PR having a higher priority.

@resetnow : it is purely frontend. It is working very well. You can already review it if you have time.

**Comment by lolodomo** [comment]
Tested with Microsoft Edge.

File: extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js [hunk 1]
```diff
@@ -1733,6 +1717,47 @@
 
 			return title;
 		};
+
+		this.updateWidget = function(widget, visibility, state, label, labelcolor, valuecolor) {
```
Comment (resetnow) [inline extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js]: would be better to make this an object:

```
function(widget, update)

// update
// {
//     state: ...,
//     label: ...,
//     And so on
// }

File: extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js [hunk 2]
```diff
@@ -1845,53 +1835,50 @@
 
 			function walkWidgets(widgets) {
 				widgets.forEach(function(widget) {
-					if (widget.item === undefined) {
+					if (widget.widgetId === undefined || smarthome.dataModel[widget.widgetId] === undefined) {
```
Comment (resetnow) [inline extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js]: linebreak between comparisons as line is too long

File: extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js [hunk 3]
```diff
@@ -1845,53 +1835,50 @@
...
-						value = _t.extractValueFromLabel(widget.label),
-						visibility = widget.visibility,
-						labelcolor = widget.labelcolor,
-						valuecolor = widget.valuecolor;
-
-					if (value === null) {
-						value = state;
-					}
+						w = smarthome.dataModel[widget.widgetId],
+						state = "NULL";
```
Comment (resetnow) [inline extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js]: perhaps `null` is better?

File: extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js (hunk 2 above)
Comment (lolodomo) [reply extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js]: Fixed

File: extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js (hunk 1 above)
Comment (lolodomo) [reply extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js]: Done

File: extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js (hunk 3 above)
Comment (lolodomo) [reply extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js]: If you look at the previous code, there was already a test again "NULL" and I think it was probably useful and I don't want to break that. If I add null here, I would have to handle null and "NULL".
I prefer a unique test against "NULL".

**Comment by lolodomo** [comment]
@resetnow: PR rebased and your comments considered. Review and merge can be restarted.

**Comment by lolodomo** [comment]
@resetnow : could you please have a final look so that it could be merged ?

I would like to switch to the next step (the PR that in in WIP stage).

File: extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js [hunk 4]
```diff
@@ -1852,53 +1856,64 @@
...
-						value = _t.extractValueFromLabel(widget.label),
-						visibility = widget.visibility,
-						labelcolor = widget.labelcolor,
-						valuecolor = widget.valuecolor;
-
-					if (value === null) {
-						value = state;
-					}
+						w = smarthome.dataModel[widget.widgetId],
+						state = "NULL",
```
Comment (resetnow) [inline extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js]: Can this ever happen?

Comment (lolodomo) [reply extensions/ui/org.eclipse.smarthome.ui.basic/web-src/smarthome.js]: Normally a state is "NULL" when the item is not yet initialized.

**Comment by resetnow** [comment]
@lolodomo could you please point out the PR where the visibility property was added to REST sitemap JSON response? I'm trying to understand whether my problems with testing are caused by incorrect launch configuration and/or build used.

**Comment by lolodomo** [comment]
The visibility property is not provided by the REST API. The REST API returns only visible widgets.

The visibility property is provided in stemap SSE.

**Comment by resetnow** [comment]
In this case what does `w.visible` refer to in `ChangeListenerLongpolling`? Have you tried this patch with default demo sitemap?

**Comment by lolodomo** [comment]
Yes tested with an enhanced demo sitemap to cover different test cases.
w.visible refer to _t.visible set in Control function.

**Pull request merged**